### DISCUSSION TITLE
censor full text line

### DISCUSSION THREAD

**Discussion by hmd-shrf** [ImageMagick/ImageMagick#2139] [Help] [answered]
I've a text image. I'd like assistance to censor the whole text-height in black in full-page-width.

I've attached some samples at the end. The original image is not necessarily in BW but the censored bars are needed in black.

Appreciate any help - thank you.

P.s working on Windows; Version: ImageMagick 7.0.10-2 Q16 x64 2020-03-22

**Orignal:**
<img width="827" alt="1" src="https://user-images.githubusercontent.com/5032598/84560391-f9eee180-ad5c-11ea-87e7-c32f73218826.png">

**Censored:**
<img width="827" alt="99111654" src="https://user-images.githubusercontent.com/5032598/84560393-02471c80-ad5d-11ea-9fb6-85c30ccf34e3.png">

**Comment by fmw42** [comment]
`convert text.png -morphology erode rectangle:3000x1 x.png`

<img width="827" alt="x" src="https://user-images.githubusercontent.com/13458218/84561148-df861b00-acfe-11ea-91d2-150e422c9c48.png">

If using ImageMagick 7, replace convert with magick.

**Reply by hmd-shrf** [reply]
Thank you @fmw42 . It works perfectly for the sample.
Unfortunately the files are I am working on has a height more than 100k pixels and when I use the above line it took more than 20 minutes to process and in the end returned an empty white document. 

What might be the cause?

**Comment by snibgo** [comment, marked as answer]
How many pixels are in your image? If you don't have enough memory, IM will use disk so it will be slow. In that case, deliberately limiting memory use may improve performance.

I don't know why you get a white image.

I think you want: where any row is entirely white, leave it white. Otherwise, make it black.

If that is what you want, we can use a more direct (and faster) method than morphology. Scale it to a single column, make all non-white pixels black, and scale it back to the original width.

For IM V7, Windows BAT script:

```
magick ^
  84560391-f9eee180-ad5c-11ea-87e7-c32f73218826.png ^
  -set option:WW %%[fx:w] ^
  -scale "1x%%[fx:h]^!" ^
  -fill Black +opaque White ^
  -scale "%%[WW]x%%[fx:h]^!" ^
  out.png
```

**Reply by hmd-shrf** [reply]
Worked!
Thank you @fmw42 and @snibgo.

**Reply by hmd-shrf** [reply]
what is the use of ! in the script, possible to guide me?

**Reply by snibgo** [reply]
"-scale NxM! will scale the image to exactly NxM pixels, without preserving the aspect ratio. See https://www.imagemagick.org/script/command-line-processing.php#geometry

**Reply by hmd-shrf** [reply]
Thank you @snibgo
